hip/test/multigrid/amgx_pgm_kernels.cpp

```diff
@@ -0,0 +1,362 @@
+}
+
+
+TEST_F(AmgxPgm, RenumberIsEquivalentToRef)
+{
+    initialize_data();
+    auto x = unfinished_agg;
+    auto d_x = d_unfinished_agg;
+    index_type num_agg;
+    index_type d_num_agg;
+
+    gko::kernels::reference::amgx_pgm::renumber(ref, agg, &num_agg);
+    gko::kernels::hip::amgx_pgm::renumber(hip, d_agg, &d_num_agg);
+
+    ASSERT_EQ(d_num_agg, num_agg);
+    GKO_ASSERT_ARRAY_EQ(d_agg, agg);
+    ASSERT_LE(num_agg, 300);
+}
+
+
+TEST_F(AmgxPgm, FindStrongestNeighborIsEquivalentToRef)
+{
+    initialize_data();
+    auto snb = strongest_neighbor;
+    auto d_snb = d_strongest_neighbor;
+
+    gko::kernels::reference::amgx_pgm::find_strongest_neighbor(
+        ref, weight_csr.get(), weight_diag.get(), agg, snb);
+    gko::kernels::hip::amgx_pgm::find_strongest_neighbor(
+        hip, d_weight_csr.get(), d_weight_diag.get(), d_agg, d_snb);
+
+    GKO_ASSERT_ARRAY_EQ(d_snb, snb);
+}
+
+
+TEST_F(AmgxPgm, AssignToExistAggIsEquivalentToRef)
+{
+    initialize_data();
+    auto x = unfinished_agg;
+    auto d_x = d_unfinished_agg;
+    auto intermediate_agg = x;
+    auto d_intermediate_agg = d_x;
+
+    gko::kernels::reference::amgx_pgm::assign_to_exist_agg(
+        ref, weight_csr.get(), weight_diag.get(), x, intermediate_agg);
+    gko::kernels::hip::amgx_pgm::assign_to_exist_agg(
+        hip, d_weight_csr.get(), d_weight_diag.get(), d_x, d_intermediate_agg);
+
+    GKO_ASSERT_ARRAY_EQ(d_x, x);
+}
+
+
+TEST_F(AmgxPgm, AssignToExistAggUnderteminsticIsEquivalentToRef)
+{
+    initialize_data();
+    auto d_x = d_unfinished_agg;
+    auto d_intermediate_agg = gko::Array<index_type>(hip, 0);
+    index_type d_num_unagg;
+
+    gko::kernels::hip::amgx_pgm::assign_to_exist_agg(
+        hip, d_weight_csr.get(), d_weight_diag.get(), d_x, d_intermediate_agg);
+    gko::kernels::hip::amgx_pgm::count_unagg(hip, d_agg, &d_num_unagg);
+
+    // only test whether all elements are aggregated.
+    GKO_ASSERT_EQ(d_num_unagg, 0);
+}
+
+
+TEST_F(AmgxPgm, GenerateMtxIsEquivalentToRef)
+{
+    initialize_data();
+    auto csr_coarse = Csr::create(ref, gko::dim<2>{n, n}, 0);
+    auto d_csr_coarse = Csr::create(hip, gko::dim<2>{n, n}, 0);
+    auto csr_temp = Csr::create(ref, gko::dim<2>{n, n},
+                                weight_csr->get_num_stored_elements());
+    auto d_csr_temp = Csr::create(hip, gko::dim<2>{n, n},
+                                  d_weight_csr->get_num_stored_elements());
+
+    gko::kernels::hip::amgx_pgm::amgx_pgm_generate(
+        hip, d_weight_csr.get(), d_agg, d_csr_coarse.get(), d_csr_temp.get());
+    gko::kernels::reference::amgx_pgm::amgx_pgm_generate(
+        ref, weight_csr.get(), agg, csr_coarse.get(), csr_temp.get());
+
+    GKO_ASSERT_MTX_NEAR(d_csr_coarse, csr_coarse, 1e-14);
+}
+
+
+TEST_F(AmgxPgm, GenerateMgLevelIsEquivalentToRef)
+{
+    initialize_data();
+    auto mg_level_factory = gko::multigrid::AmgxPgm<double, int>::build()
+                                .with_deterministic(true)
```

Comment:
Maybe also a test with `false` here? Would that be possible or are the results way too different that you cannot find a tolerance for which it works?

Reply:
when it is false, the dimension may be different. Thus, I can only test whether it is valid in AssignToExistAggUnderteminsticIsEquivalentToRef